ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java

```diff
@@ -0,0 +1,118 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The storage manager that does book keeping for all of the partitions handled by this vcr node.
+ */
+public class CloudStorageManager implements StoreManager {
+  private final ConcurrentMap<PartitionId, CloudBlobStore> partitionToStore;
+  private final VerifiableProperties properties;
+  private final CloudDestination cloudDestination;
+  private final VcrMetrics vcrMetrics;
+  private final ClusterMap clusterMap;
+  private static final Logger logger = LoggerFactory.getLogger(CloudStorageManager.class);
+
+  public CloudStorageManager(VerifiableProperties properties, VcrMetrics vcrMetrics, CloudDestination cloudDestination,
+      ClusterMap clusterMap) {
+    partitionToStore = new ConcurrentHashMap<>();
+    this.properties = properties;
+    this.cloudDestination = cloudDestination;
+    this.vcrMetrics = vcrMetrics;
+    this.clusterMap = clusterMap;
+  }
+
+  @Override
+  public boolean addBlobStore(ReplicaId replica) {
+    CloudBlobStore cloudStore = partitionToStore.getOrDefault(replica.getPartitionId(), null);
+    if (cloudStore == null) {
+      cloudStore = new CloudBlobStore(properties, replica.getPartitionId(), cloudDestination, clusterMap, vcrMetrics);
+      partitionToStore.put(replica.getPartitionId(), cloudStore);
+    }
+    return startBlobStore(replica.getPartitionId());
+  }
+
+  @Override
+  public boolean shutdownBlobStore(PartitionId id) {
+    CloudBlobStore blobStore = partitionToStore.getOrDefault(id, null);
+    if (blobStore == null) {
+      return false;
+    }
+    blobStore.shutdown();
+    return true;
+  }
+
+  @Override
+  public Store getStore(PartitionId id) {
+    return partitionToStore.getOrDefault(id, null);
+  }
+
+  @Override
+  public boolean scheduleNextForCompaction(PartitionId id) {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public boolean startBlobStore(PartitionId id) {
+    CloudBlobStore cloudStore = partitionToStore.getOrDefault(id, null);
+    if (cloudStore == null) {
+      return false;
+    }
+    try {
+      cloudStore.start();
+    } catch (StoreException e) {
+      logger.error("Can't start CloudStore " + cloudStore, e);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public List<PartitionId> setBlobStoreStoppedState(List<PartitionId> partitionIds, boolean markStop) {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public boolean removeBlobStore(PartitionId id) {
+    return partitionToStore.remove(id) != null;
+  }
+
+  @Override
+  public ServerErrorCode checkLocalPartitionStatus(PartitionId partition, ReplicaId localReplica) {
+    if (getStore(partition) == null) {
+      return ServerErrorCode.Replica_Unavailable;
```

Comment:
Should this return Replica_Unavailable if a store is still in the map but has been shutdown?

In fact, the behavior of `getStore` in `CloudStorageManager` differs slightly from `DiskManager`. In `DiskManager` we do the following check:
```java
    BlobStore store = stores.get(id);
    return (running && store != null && store.isStarted()) ? store : null;
```

Reply:
StorageManager::checklocalPartitionStatus returns one of four error codes : Disk_Unavailable, Replica_Unavailable, Partition_Unknown and No_Error.
For CloudStoreManager, Disk_Unavailable doesn't make sense. Also, note that the only state that determines whether a vcr can serve a replica, is the state it gets from clustermanager. There is no disk for it to verify against. So I guess we either need to throw one of Replica_Unavailable or PArtition_Unknown in this case. Or we can also introduce a new state called Partition_Unassigned, as truly, this only means that the partition isnt assigned to this VCR, and we dont really know if Replica or Partition is unavailble. Let me know what you think of this.

Also fixed getStore.

Reply:
fixed as discussed. we will return Partition_Unknown if the partition isnt assigned to vcr. We will return Replica_Unavailable if partition is stopped (blobstore is stopped).